### PR TITLE
Start xstate implementation

### PR DESCRIPTION
**Pull request by muhajirdev**
TODO

- [x] remaining count of the repeat butotn
- [x] change reciter from UI
- [x] fix broken localstorage persistance
- [x] play audio without changing reciter (need to update xstate implementation)
- [x] remove unused selectors  & actions from redux
- [x] persist reciter data & playback rate to DB
- [x] Inifinite repeat settings
- [x] Load reciterId from database instead of localstorage and save to database via (onSettingsChangeXstate)
- [x] Optimize rerendering, replace useActor with useSelector
- [x] update typing file, use the existing one
- [x] fetch data using existing fetcher -> the implication is we'll use verseTiming instaed of verse_timing, verseKey instead of verse_key

### CODE REVIEW

**Comment by mmahalwy** [comment]
#### 📝 Changed routes:


| [**…reciterId\]/\[chapterId\]**](https://quran-opx6hyu1n-qurancom.vercel.app/reciters/[reciterId]/[chapterId]) |
|:-:|
|<a href="https://quran-opx6hyu1n-qurancom.vercel.app/reciters/[reciterId]/[chapterId]">Screenshot of /reciters/[reciterId]/[chapterId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/lqKOz0zeyaNAUVvH5QunkPwnbqS8UIZX3vgVenlvUbmCq4yeAxsoKD3o-VsYOeHvqhmbdk39-Puh6sGVnUHcUw.png">(view full size)</a> |







Commit 3ff1835c56181b3c85b78536dd603324450426a1 (https://quran-opx6hyu1n-qurancom.vercel.app).

File: src/xstate/constants.ts [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+/* eslint-disable @typescript-eslint/naming-convention */
+export const SURAH_AYAHS_COUNT = [
+  7, 286, 200, 176, 120, 165, 206, 75, 129, 109, 123, 111, 43, 52, 99, 128, 111, 110, 98, 135, 112,
+  78, 118, 64, 77, 227, 93, 88, 69, 60, 34, 30, 73, 54, 45, 83, 182, 88, 75, 85, 54, 53, 89, 59, 37,
+  35, 38, 29, 18, 45, 60, 49, 62, 55, 78, 96, 29, 22, 24, 13, 14, 11, 11, 18, 12, 12, 30, 52, 52,
+  44, 28, 28, 20, 56, 40, 31, 50, 40, 46, 42, 29, 19, 36, 25, 22, 17, 19, 26, 30, 20, 15, 21, 11, 8,
+  8, 19, 5, 8, 8, 11, 11, 8, 3, 9, 5, 4, 7, 3, 6, 3, 5, 4, 5, 6,
+];
+
+export const RECITERS = [
```
Comment (osamasayed) [inline src/xstate/constants.ts:11]: This should not be hard-coded.

File: src/xstate/constants.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
+/* eslint-disable max-lines */
+/* eslint-disable @typescript-eslint/naming-convention */
+export const SURAH_AYAHS_COUNT = [
```
Comment (osamasayed) [inline src/xstate/constants.ts:3]: We already have `versesCount` inside `data/chapters/en.json` file.

File: src/xstate/actors/verseCycle/verseCycleMachine.typegen.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  };
+  eventsCausingServices: {};
+  eventsCausingGuards: {
+    verseEnded: "TIMESTAMP_UPDATED";
+    verseRepeatOnProgress: "";
+  };
+  eventsCausingDelays: {};
+  matchesStates: "finished" | "inProgress" | "verseEnded";
+  tags: never;
+}
```
Comment (osamasayed) [inline src/xstate/actors/verseCycle/verseCycleMachine.typegen.ts:29]: This file should not be committed

File: src/xstate/actors/repeatMachine/types/VerseTiming.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+type VerseTiming = {
+  verse_key: string;
+  timestamp_from: number;
+  timestamp_to: number;
+  duration: number;
+  segments: Segment[];
+};
+
+export default VerseTiming;
```
Comment (osamasayed) [inline src/xstate/actors/repeatMachine/types/VerseTiming.ts:12]: We already have `types/VerseTiming.ts` file.

File: src/xstate/actors/repeatMachine/types/Segment.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+type Segment = [number, number, number];
+
+export default Segment;
```
Comment (osamasayed) [inline src/xstate/actors/repeatMachine/types/Segment.ts:3]: We already have `types/Segment.ts`.

File: src/xstate/actors/audioPlayer/types/services/AudioData.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    duration: number;
+    segments: number[][];
+  }[];
+};
+
+type AudioData = {
+  audio_files: AudioFile[];
+};
+
+export default AudioData;
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/types/services/AudioData.ts:22]: We already have `types/AudioData.ts`.

File: src/xstate/Radio/types/StationInfo.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+type StationInfo = {
+  title: string;
+  description: string;
+};
+
+export default StationInfo;
```
Comment (osamasayed) [inline src/xstate/Radio/types/StationInfo.ts:6]: We already have this inside `src/components/Radio/types.ts`

File: src/xstate/Radio/types/CuratedStation.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import AudioTrack from './AudioTrack';
+
+type CuratedStation = {
+  audioTracks: AudioTrack[];
+  title: string;
+  description: string;
+  bannerImgSrc?: string;
+};
+
+export default CuratedStation;
```
Comment (osamasayed) [inline src/xstate/Radio/types/CuratedStation.ts:10]: We already have this inside `src/components/Radio/types.ts`

File: src/xstate/Radio/types/AudioTrack.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+type AudioTrack = {
+  reciterId: string;
+  surah: string;
+};
+
+export default AudioTrack;
```
Comment (osamasayed) [inline src/xstate/Radio/types/AudioTrack.ts:6]: We already have this inside `src/components/Radio/types.ts`

File: src/xstate/AudioPlayerMachineContext.tsx [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+      if (state.event.type !== 'UPDATE_TIMING') console.log(state.value, state.event);
+      persistXstateToLocalStorage({ playbackRate, reciterId });
+    },
+  );
+
+  useEffect(() => {
+    if (isClient && loggedIn) {
+      getUserPreferences(locale).then((preferences) => {
+        const playbackRate =
+          preferences[PreferenceGroup.AUDIO].playbackRate ||
```
Comment (osamasayed) [inline src/xstate/AudioPlayerMachineContext.tsx:57]: This should be `preferences[PreferenceGroup.AUDIO]?.playbackRate` since it can be `undefined`.

File: src/xstate/AudioPlayerMachineContext.tsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    {
+      context: {
+        ...audioPlayerMachine.initialState.context,
+        ...defaultLocaleContext,
+        ...initialXstateContext,
+      },
+    },
+    (state) => {
+      const { playbackRate, reciterId } = state.context;
+      if (state.event.type !== 'UPDATE_TIMING') console.log(state.value, state.event);
```
Comment (osamasayed) [inline src/xstate/AudioPlayerMachineContext.tsx:48]: This should be removed.

File: src/xstate/AudioPlayerMachineContext.tsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      persistXstateToLocalStorage({ playbackRate, reciterId });
+    },
+  );
+
+  useEffect(() => {
+    if (isClient && loggedIn) {
+      getUserPreferences(locale).then((preferences) => {
```
Comment (osamasayed) [inline src/xstate/AudioPlayerMachineContext.tsx:55]: this will cause 2 requests to `getUserPreferences` API since we already call it inside `src/redux/Provider.tsx`.`onBeforeLift`

File: src/xstate/AudioPlayerMachineContext.tsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      context: {
+        ...audioPlayerMachine.initialState.context,
+        ...defaultLocaleContext,
+        ...initialXstateContext,
+      },
+    },
+    (state) => {
+      const { playbackRate, reciterId } = state.context;
+      if (state.event.type !== 'UPDATE_TIMING') console.log(state.value, state.event);
+      persistXstateToLocalStorage({ playbackRate, reciterId });
```
Comment (osamasayed) [inline src/xstate/AudioPlayerMachineContext.tsx:49]: Shouldn't this be called only when events that change the reciter or playbackRate are triggered?

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 14]
```diff
@@ -0,0 +1,1017 @@
...
+  return {
+    ...data,
+    ...event,
+  };
+};
+
+const getMediaSessionMetaData = async (context: AudioPlayerContext) => {
+  const reciterName = RECITERS.find((reciter) => reciter.id === context.reciterId).name;
+  return new MediaMetadata({
+    title: `Surah ${context.audioData.chapterId}`,
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:115]: This is in a followup but I think we should localize this

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 15]
```diff
@@ -0,0 +1,1017 @@
...
+        SET_INITIAL_CONTEXT: {
+          actions: 'setInitialContext',
+        },
+      },
+    },
+    {
+      actions: {
+        test: () => {
+          alert('a');
+        },
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:686]: This should be removed

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 16]
```diff
@@ -0,0 +1,1017 @@
...
+        isNotFirstVerse: (context) => context.ayahNumber !== 1,
+        isDifferentSurah: (context, event) => context.surah !== event.surah,
+        isSameSurah: (context, event) => context.surah === event.surah,
+        isRepeatActive: (context) => !!context.repeatActor,
+        isUsingCustomReciterId: (context, event) => !!event.reciterId,
+      },
+      services: {
+        fetchReciter: (context) => executeFetchReciter(context),
+        fetchCustomReciter: (context, event) => {
+          console.log(event);
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:955]: This should be removed

File: src/components/ThirdPartyScripts/ThirdPartyScripts.tsx [hunk 17]
```diff
@@ -1,13 +1,13 @@
...
-import Script from 'next/script';
+// import Script from 'next/script';
 
 import GoogleAnalyticsScript from './GoogleAnalyticsScript';
 
 const ThirdPartyScripts = () => (
   <>
     <GoogleAnalyticsScript />
-    <Script
+    {/* <Script
```
Comment (osamasayed) [inline src/components/ThirdPartyScripts/ThirdPartyScripts.tsx:10]: I believe this is un-intentional

File: src/utils/query-params.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
+import { areArraysEqual } from './array';
```
Comment (osamasayed) [inline src/utils/query-params.ts:1]: Can be in a followup but I believe this util needs some tests

File: src/utils/query-params.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+export enum ValueType {
```
Comment (osamasayed) [inline src/utils/query-params.ts:3]: nit: `ValueType` should be renamed to something more specific e.g. `QueryParamValueType`

File: src/utils/query-params.ts [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+  Number = 'Number',
+  ArrayOfNumbers = 'ArrayOfNumbers',
+  ArrayOfStrings = 'ArrayOfStrings',
+}
+
+const paramStringValueParser = {
+  [ValueType.ArrayOfNumbers]: (paramStringValue: string) =>
+    paramStringValue.split(',').map((stringValue) => Number(stringValue)),
+  [ValueType.String]: (paramStringValue: string) =>
+    paramStringValue.split(',').map((stringValue) => Number(stringValue)),
```
Comment (osamasayed) [inline src/utils/query-params.ts:14]: Why do we need to split for a single `String` value?

File: src/utils/query-params.ts [hunk 21]
```diff
@@ -0,0 +1,30 @@
+import { areArraysEqual } from './array';
+
+export enum ValueType {
+  String = 'String',
+  Number = 'Number',
+  ArrayOfNumbers = 'ArrayOfNumbers',
+  ArrayOfStrings = 'ArrayOfStrings',
+}
+
+const paramStringValueParser = {
```
Comment (osamasayed) [inline src/utils/query-params.ts:10]: nit : `paramValueParser`

File: src/hooks/auth/usePersistPreferenceGroup.ts [hunk 22]
```diff
@@ -69,6 +78,55 @@ const usePersistPreferenceGroup = (): PersistPreferences => {
           callback();
         }
       },
+      // TODO: add onSetttingsChange xstate
```
Comment (osamasayed) [inline src/hooks/auth/usePersistPreferenceGroup.ts:81]: nit: TODO should be removed

File: src/hooks/auth/usePersistPreferenceGroup.ts [hunk 23]
```diff
@@ -69,6 +78,55 @@ const usePersistPreferenceGroup = (): PersistPreferences => {
...
+                    },
+                  },
+                ],
+              });
+            })
+            .finally(() => {
+              setIsLoading(false);
+            });
+        } else {
+          dispatch(action);
```
Comment (osamasayed) [inline src/hooks/auth/usePersistPreferenceGroup.ts:127]: this should be `action()`.

File: src/xstate/AudioPlayerMachineContext.tsx (hunk 12 above)
Comment (muhajirdev) [reply src/xstate/AudioPlayerMachineContext.tsx:55]: yes, you're right. I am still thinking if there's a good way to fix this issue

File: src/components/Radio/ReciterStationList.tsx [hunk 24]
```diff
@@ -57,10 +39,14 @@ const ReciterStationList = ({ reciters }: ReciterStationListProps) => {
...
+            audioService.send({
+              type: 'PLAY_RADIO',
+              stationType: StationType.Reciter,
+              stationId: reciter.id,
+            });
           };
-        if (isSelectedStation && isAudioPlaying) onClick = () => triggerPauseAudio();
-        if (isSelectedStation && !isAudioPlaying) onClick = () => triggerPlayAudio(playbackRate);
+        if (isSelectedStation && isAudioPlaying) onClick = () => audioService.send('TOGGLE');
+        if (isSelectedStation && !isAudioPlaying) onClick = () => audioService.send('TOGGLE');
```
Comment (osamasayed) [inline src/components/Radio/ReciterStationList.tsx:49]: this should be simplified, we are unnecessarily checking `isAudioPlaying` while if it's true or false the `onClick` value is the same

File: src/components/AudioPlayer/SeekButton.tsx [hunk 25]
```diff
@@ -36,60 +18,22 @@ type SeekButtonProps = {
...
-    !chapterAudioData ||
-    (type === SeekButtonType.PrevAyah && highlightedVerse <= 1) ||
-    (type === SeekButtonType.NextAyah && highlightedVerse >= chapterData?.versesCount);
-
   return (
     <Button
       tooltip={type === SeekButtonType.PrevAyah ? t('previous-ayah') : t('next-ayah')}
       variant={ButtonVariant.Ghost}
       shape={ButtonShape.Circle}
-      isDisabled={isDisabled}
```
Comment (osamasayed) [inline src/components/AudioPlayer/SeekButton.tsx:92]: Don't we need to disable the button when we reach either the first or last verse of the current Surah?

File: src/components/HomePage/PlayRadioButton.tsx [hunk 26]
```diff
@@ -65,14 +89,8 @@ const PlayRadioButton = () => {
...
-
-        {shouldShowStationName && (
-          <div className={styles.stationInfo}>
-            <span className={styles.stationTitle}>{stationInfo.title}</span>{' '}
-            <Link href="/radio" className={styles.editStationButton}>
-              ({t('change')})
-            </Link>
-          </div>
+        {audioService.getSnapshot().context.radioActor && (
+          <RadioInformation radioActor={audioService.getSnapshot().context.radioActor} />
```
Comment (osamasayed) [inline src/components/HomePage/PlayRadioButton.tsx:93]: `audioService.getSnapshot().context.radioActor` is being repeated twice, we should put it in a variable

File: src/components/HomePage/PlayRadioButton.tsx [hunk 27]
```diff
@@ -1,55 +1,79 @@
...
 
 import styles from './PlayRadioButton.module.scss';
 
-import { playFrom, selectIsPlaying, selectIsRadioMode } from 'src/redux/slices/AudioPlayer/state';
-import { selectRadioStation } from 'src/redux/slices/radio';
 import { logEvent } from 'src/utils/eventLogger';
+import { AudioPlayerMachineContext } from 'src/xstate/AudioPlayerMachineContext';
+import useCurrentStationInfo from 'src/xstate/Radio/useCurrentStationInfo';
+
+const RadioInformation = ({ radioActor }) => {
```
Comment (osamasayed) [inline src/components/HomePage/PlayRadioButton.tsx:21]: This should be put in a separate file and `/* eslint-disable react/no-multi-comp */` line should be removed

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 28]
```diff
@@ -0,0 +1,1027 @@
...
+ * - timestampFrom = 10, timestampTo = 20, currentTime = 10 should return true
+ * - timestampFrom = 10, timestampTo = 20, currentTime = 11 should return true
+ * - timestampFrom = 10, timestampTo = 20, currentTime = 20 should return false
+ *
+ * @param {number} currentTime
+ * @param {number} timestampFrom
+ * @param {number} timestampTo
+ * @returns {boolean} isWithinRange
+ */
+const isCurrentTimeInRange = (currentTime: number, timestampFrom: number, timestampTo: number) =>
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:35]: I suggest this being moved to a helper file since the file is very bloated already and hard to follow.

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 29]
```diff
@@ -0,0 +1,1027 @@
...
+      ayah.timestampFrom,
+      ayah.timestampTo,
+    );
+    return isAyahBeingRecited;
+  });
+
+  return activeVerseTiming;
+};
+
+export const getActiveWordLocation = (activeVerseTiming: VerseTiming, currentTime: number) => {
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:64]: Ditto: I suggest this being moved to a helper file since the file is very bloated already and hard to follow.

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 30]
```diff
@@ -0,0 +1,1027 @@
...
+  const activeAudioSegment = activeVerseTiming.segments.find((segment) => {
+    const [, timestampFrom, timestampTo] = segment; // the structure of the segment is: [wordLocation, timestampFrom, timestampTo]
+    return isCurrentTimeInRange(currentTime, timestampFrom, timestampTo);
+  });
+
+  const wordLocation = activeAudioSegment ? activeAudioSegment[0] : 0;
+  return wordLocation;
+};
+
+const getTimingSegment = (verseTiming: VerseTiming, wordPosition: number) =>
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:74]: Ditto: I suggest this being moved to a helper file since the file is very bloated already and hard to follow.

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 31]
```diff
@@ -0,0 +1,1027 @@
...
+  verseTiming.segments.find(([location]) => wordPosition === location);
+
+export const getWordTimeSegment = (verseTimings: VerseTiming[], word: Word) => {
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:77]: Ditto: I suggest this being moved to a helper file since the file is very bloated already and hard to follow.

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 32]
```diff
@@ -0,0 +1,1027 @@
...
+  const verseTiming = verseTimings.find((timing) => timing.verseKey === word.verseKey);
+  if (!verseTiming) return null;
+  const segment = getTimingSegment(verseTiming, word.position);
+  if (segment) return [segment[1], segment[2]];
+  return null;
+};
+
+const getActiveAyahNumber = (activeVerseTiming: VerseTiming) => {
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:85]: Ditto: I suggest this being moved to a helper file since the file is very bloated already and hard to follow.

File: src/components/QuranReader/PlayChapterAudioButton.tsx [hunk 33]
```diff
@@ -1,58 +1,54 @@
...
+
+    // if surah is being played, toggle it
+    if (Number(chapterId) === surah) {
+      audioService.send('TOGGLE');
     } else {
-      dispatch(playFrom({ chapterId, reciterId, timestamp: 0 }));
+      audioService.send({
+        type: 'PLAY_SURAH',
+        surah: chapterId,
+      });
```
Comment (osamasayed) [inline src/components/QuranReader/PlayChapterAudioButton.tsx:43]: Do you think this condition should be checked here or inside `PLAY_SURAH`? I am leaning towards inside the xstate's event handler

File: src/components/QuranReader/PlayChapterAudioButton.tsx [hunk 34]
```diff
@@ -1,58 +1,54 @@
...
-  const { value: reciterId }: { value: number } = useGetQueryParamOrReduxValue(QueryParam.Reciter);
   const play = () => {
     logButtonClick('chapter_header_play_audio');
-    dispatch(exitRepeatMode());
-    if (currentAudioData?.chapterId === chapterId && !isRadioMode) {
-      triggerPlayAudio(playbackRate);
+    const { surah } = audioService.getSnapshot().context;
+
+    // if surah is being played, toggle it
+    if (Number(chapterId) === surah) {
```
Comment (osamasayed) [inline src/components/QuranReader/PlayChapterAudioButton.tsx:37]: But if it's being played already and the user clicks play chapter again, it will be paused this way if we are dispatching `TOGGLE` event.

File: src/components/QuranReader/ReadingView/Line.tsx [hunk 35]
```diff
@@ -27,9 +28,18 @@ export type LineProps = {
...
+    const { surah, ayahNumber } = state.context;
+    const verseKeys = words.map((word) => word.verseKey);
+    return verseKeys.includes(`${surah}:${ayahNumber}`);
+  });
+
   const [scrollToSelectedItem, selectedItemRef]: [() => void, RefObject<HTMLDivElement>] =
     useScroll(SMOOTH_SCROLL_TO_CENTER);
+
+  // export c  const [scrollToSelectedItem, selectedItemRef]: [() => void, RefObject<HTMLDivElement>] =
+  useScroll(SMOOTH_SCROLL_TO_CENTER);
```
Comment (osamasayed) [inline src/components/QuranReader/ReadingView/Line.tsx:42]: I believe this was left by mistake

File: src/components/AudioPlayer/RepeatButton.tsx [hunk 36]
```diff
@@ -11,23 +12,28 @@ import { RepetitionMode } from './RepeatAudioModal/SelectRepetitionMode';
...
-  selectIsInRepeatMode,
-  selectRemainingRangeRepeatCount,
-  selectAudioData,
-} from 'src/redux/slices/AudioPlayer/state';
 import { logButtonClick } from 'src/utils/eventLogger';
 import { toLocalizedNumber } from 'src/utils/locale';
+import { AudioPlayerMachineContext } from 'src/xstate/AudioPlayerMachineContext';
 
-const RepeatAudioButton = () => {
+const RemainingRangeCount = ({ rangeActor }) => {
```
Comment (osamasayed) [inline src/components/AudioPlayer/RepeatButton.tsx:19]: Should be split into a separate component

File: src/components/AudioPlayer/RepeatButton.tsx [hunk 37]
```diff
@@ -11,23 +12,28 @@ import { RepetitionMode } from './RepeatAudioModal/SelectRepetitionMode';
...
   const [isModalOpen, setIsModalOpen] = useState(false);
-  const audioData = useSelector(selectAudioData, shallowEqual);
-  const isInRepeatMode = useSelector(selectIsInRepeatMode);
-  const remainingRangeRepeatCount = toLocalizedNumber(
-    useSelector(selectRemainingRangeRepeatCount),
-    lang,
-  );
+  const isInRepeatMode = useSelector(audioService, (state) => !!state.context.repeatActor);
+  const currentSurah = useSelector(audioService, (state) => state.context.surah);
+  const repeatActor = useSelector(audioService, (state) => state.context.repeatActor);
```
Comment (osamasayed) [inline src/components/AudioPlayer/RepeatButton.tsx:36]: This can be simplified as following:

```
  const currentSurah = useSelector(audioService, (state) => state.context.surah);
  const repeatActor = useSelector(audioService, (state) => state.context.repeatActor);
  const isInRepeatMode = !!repeatActor;
```

File: src/components/AudioPlayer/RepeatButton.tsx [hunk 38]
```diff
@@ -36,25 +42,30 @@ const RepeatAudioButton = () => {
...
+        onClose={() => setIsModalOpen(false)}
+      />
 
       <Wrapper
         shouldWrap={isInRepeatMode}
-        wrapper={(children) => <Badge content={remainingRangeRepeatCount}>{children}</Badge>}
+        wrapper={(children) => (
+          <Badge
+            content={
+              repeatActor && (
```
Comment (osamasayed) [inline src/components/AudioPlayer/RepeatButton.tsx:57]: `isInRepeatMode && (`

File: src/components/dls/QuranWord/QuranWord.tsx [hunk 39]
```diff
@@ -121,13 +111,25 @@ const QuranWord = ({
...
+      const currentState = audioService.getSnapshot();
+      const isPlaying = currentState.matches('VISIBLE.AUDIO_PLAYER_INITIATED.PLAYING');
+      const currentSurah = getChapterNumberFromKey(word.verseKey);
+      const isSameSurah = currentState.context.surah === Number(currentSurah);
+      const shouldSeekTo = isPlaying && isSameSurah;
+      if (shouldSeekTo) {
+        const wordSegment = getWordTimeSegment(currentState.context.audioData.verseTimings, word);
+        if (!wordSegment) return;
+        const [startTime] = wordSegment;
+        audioService.send({ type: 'SEEK_TO', timestamp: startTime / 1000 });
```
Comment (osamasayed) [inline src/components/dls/QuranWord/QuranWord.tsx:125]: We have a util `milliSecondsToSeconds` the we should use.

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 40]
```diff
@@ -0,0 +1,1027 @@
...
+        setCurrentSurahAndAyahAsNewSurahAndAyah: assign({
+          surah: (context) => context.newSurah,
+          ayahNumber: (context) => context.newAyahNumber,
+        }),
+        setAudioRef: assign({
+          audioPlayer: (context, event) => event.audioPlayerRef,
+        }),
+        setAudioData: assign({
+          duration: (context, event: any) => {
+            return event.data.audioFiles[0].duration / 1000;
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:788]: We have a util `milliSecondsToSeconds` the we should use.

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 41]
```diff
@@ -0,0 +1,1027 @@
...
+            duration,
+            shouldPlayFromRandomTimeStamp,
+          } = context;
+          if (shouldPlayFromRandomTimeStamp) {
+            const randomTimestamp = random(0, duration);
+            context.audioPlayer.currentTime = randomTimestamp;
+          } else {
+            const ayahTimestamps = verseTimings[ayahNumber - 1];
+            const { timestampFrom } = ayahTimestamps;
+            context.audioPlayer.currentTime = timestampFrom / 1000;
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts:812]: We have a util `milliSecondsToSeconds` the we should use.

File: src/components/Navbar/SettingsDrawer/ReciterSelectionBody.tsx [hunk 42]
```diff
@@ -94,18 +93,19 @@ const SettingsReciter = () => {
...
-                .sort((a, b) => (a.name > b.name ? 1 : -1))
+                .sort((a, b) => (a.name + a.id > b.name + b.id ? 1 : -1))
                 .map((reciter) => {
                   const reciterId = reciter.id.toString();
                   return (
                     <div className={styles.reciter} key={reciterId}>
                       <RadioGroup.Item value={reciterId} id={reciterId} />
 
                       <label htmlFor={reciterId} className={styles.reciterLabel}>
+                        {reciterId}
```
Comment (osamasayed) [inline src/components/Navbar/SettingsDrawer/ReciterSelectionBody.tsx:108]: This should be removed

**Comment by mmahalwy** [comment]
#### 📝 Changed routes:


| [**…reciterId\]/\[chapterId\]**](https://quran-iyooxp0oh-qurancom.vercel.app/reciters/[reciterId]/[chapterId]) |
|:-:|
|<a href="https://quran-iyooxp0oh-qurancom.vercel.app/reciters/[reciterId]/[chapterId]">Screenshot of /reciters/[reciterId]/[chapterId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/nRjrLp9ZvS1RzcVXTAsGjsL0ZnRSsgz77TsJud3wZrfiDPL4XvSd5AJZhHFpt7L2t67T9vbV4HnSpPnsozt7LA.png">(view full size)</a> |







Commit 7127a716b9015ae2b51a717ee2d865d88751558e (https://quran-iyooxp0oh-qurancom.vercel.app).

**Comment by mmahalwy** [comment]
#### 📝 Changed routes:


| [**…reciterId\]/\[chapterId\]**](https://quran-pz4ukjra5-qurancom.vercel.app/reciters/[reciterId]/[chapterId]) |
|:-:|
|<a href="https://quran-pz4ukjra5-qurancom.vercel.app/reciters/[reciterId]/[chapterId]">Screenshot of /reciters/[reciterId]/[chapterId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/r5G2_kTRLcpV1iCgsYw8BIfK5m0tHhb9tT8yf2Po3rM292AY0DU81ONx2E-ybFQfsYF-oTDB8Ls7GDhCuTKT2A.png">(view full size)</a> |







Commit 2f2f1e0202be066b7c0647cff1fb765ecc404a07 (https://quran-pz4ukjra5-qurancom.vercel.app).

File: src/components/HomePage/PlayRadioButton.tsx [hunk 43]
```diff
@@ -1,55 +1,64 @@
...
   };
 
+  const { radioActor } = audioService.getSnapshot().context;
   return (
     <div className={styles.container}>
       <div className={styles.playRadioSection}>
         {isAudioPlaying && isRadioMode ? (
           <Button
-            prefix={<PauseIcon />}
+            prefix={isLoading ? <Spinner /> : <PauseIcon />}
```
Comment (osamasayed) [inline src/components/HomePage/PlayRadioButton.tsx:61]: Since we are already inside `isAudioPlaying`, `isLoading` will never be `true`.

**Comment by mmahalwy** [comment]
#### 📝 Changed routes:


| [**…reciterId\]/\[chapterId\]**](https://quran-ollcig6db-qurancom.vercel.app/reciters/[reciterId]/[chapterId]) |
|:-:|
|<a href="https://quran-ollcig6db-qurancom.vercel.app/reciters/[reciterId]/[chapterId]">Screenshot of /reciters/[reciterId]/[chapterId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/5Ca6T4HABoWGpRdV5uIaD1txgTlwXLGIERjTRwK1pkj3CbFyNR4rClnIDZNEhqpCHh2kr4Bv-ROG-t1V08hPXQ.png">(view full size)</a> |







Commit e4d65860629605ef178dc0112e59439c2d381302 (https://quran-ollcig6db-qurancom.vercel.app).

File: src/xstate/actors/audioPlayer/audioPlayerMachine.ts [hunk 44]
```diff
@@ -1009,6 +1048,15 @@ export const audioPlayerMachine =
         },
         isRepeatActive: (context) => !!context.repeatActor,
         isUsingCustomReciterId: (context, event) => !!event.reciterId,
+        isAudioAlmostEnded: (context) => {
+          const { currentTime } = context.audioPlayer;
+          const duration = milliSecondsToSeconds(context.audioData.duration);
+          const durationWithTolerancePeriod = duration - 3;
+
+          console.log('almost ended', currentTime > durationWithTolerancePeriod);
```
Comment (osamasayed) [inline src/xstate/actors/audioPlayer/audioPlayerMachine.ts]: I believe this was forgotten :)

File: src/components/HomePage/PlayRadioButton.tsx (hunk 43 above)
Comment (muhajirdev) [reply src/components/HomePage/PlayRadioButton.tsx:61]: maybe it still can? It can go to PLAYING.LOADING and PLAYING.ACTIVE?

**Comment by mmahalwy** [comment]
#### 📝 Changed routes:


| [**…reciterId\]/\[chapterId\]**](https://quran-f0vmrwfbg-qurancom.vercel.app/reciters/[reciterId]/[chapterId]) |
|:-:|
|<a href="https://quran-f0vmrwfbg-qurancom.vercel.app/reciters/[reciterId]/[chapterId]">Screenshot of /reciters/[reciterId]/[chapterId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/t84EsAPUylVzDhNO7e8iEPJJ3UYep3l6nwJ1oAKAR38tMVUH5lVWougs8sO8SU0x5fjNVFWrB5u_riKargxBFg.png">(view full size)</a> |







Commit 3bf7b5cccb23f3988ecbda0d5a99b6e8afba528f (https://quran-f0vmrwfbg-qurancom.vercel.app).

**Review by osamasayed — APPROVED** [review]

**Comment by mmahalwy** [comment]
#### 📝 Changed routes:


| [**…reciterId\]/\[chapterId\]**](https://quran-foa890odq-qurancom.vercel.app/reciters/[reciterId]/[chapterId]) |
|:-:|
|<a href="https://quran-foa890odq-qurancom.vercel.app/reciters/[reciterId]/[chapterId]">Screenshot of /reciters/[reciterId]/[chapterId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/crEjXxthnIs3Dkyx8U8m-S-svDRprzmhvi27G38m5zkevw9KfdCA1IJqAP29uv1z1CoTl1HTi1s6uMnte4_M1g.png">(view full size)</a> |







Commit 2c9daf3f9c908f49107ce18c4a5d32cbe73b79e6 (https://quran-foa890odq-qurancom.vercel.app).
